In [2]:
import numpy as np
import pandas as pd

pd.set_option("display.precision", 2)

We'll demonstrate the main methods in action by analyzing a dataset on the churn rate of telecom operator clients. Let's read the data (using thr read_csv), and take a look at the first 5 lines using the head method:

In [3]:
df = pd.read_csv(r"D:\University\3-2\3211\Telecom Churn.csv")
df.head()

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   
1    OH             107        415     371-7191                 no   
2    NJ             137        415     358-1921                 no   
3    OH              84        408     375-9999                yes   
4    OK              75        415     330-6626                yes   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  

[5 rows x 21 columns]

Recall that each row corresponds to one client, an instance, and columns are features of this instance.

Let’s have a look at data dimensionality, feature names, and feature types.

In [11]:
print(df.shape)

(3333, 21)


From the output, we can see that the table contains 3333 rows and 20 columns.

Now let's try printing out column names using columns:

In [12]:
print(df.columns)

Index(['state', 'account length', 'area code', 'phone number',
       'international plan', 'voice mail plan', 'number vmail messages',
       'total day minutes', 'total day calls', 'total day charge',
       'total eve minutes', 'total eve calls', 'total eve charge',
       'total night minutes', 'total night calls', 'total night charge',
       'total intl minutes', 'total intl calls', 'total intl charge',
       'customer service calls', 'churn'],
      dtype='object')


We can use the info() method to output some general information about the dataframe: 

In [13]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   3333 non-null   object 
 1   account length          3333 non-null   int64  
 2   area code               3333 non-null   int64  
 3   phone number            3333 non-null   object 
 4   international plan      3333 non-null   object 
 5   voice mail plan         3333 non-null   object 
 6   number vmail messages   3333 non-null   int64  
 7   total day minutes       3333 non-null   float64
 8   total day calls         3333 non-null   int64  
 9   total day charge        3333 non-null   float64
 10  total eve minutes       3333 non-null   float64
 11  total eve calls         3333 non-null   int64  
 12  total eve charge        3333 non-null   float64
 13  total night minutes     3333 non-null   float64
 14  total night calls       3333 non-null   

bool, int64, float64 and object are the data types of our features. We see that one feature is logical (bool), 3 features are of type object, and 16 features are numeric. With this same method, we can easily see if there are any missing values. Here, there are none because each column contains 3333 observations, the same number of rows we saw before with shape.

We can change the column type with the astype method. Let's apply this method to the Churn feature to convert it into int64:

In [14]:
df["churn"] = df["churn"].astype("int64")

The describe method shows basic statistical characteristics of each numerical feature (int64 and float64 types): number of non-missing values, mean, standard deviation, range, median, 0.25 and 0.75 quartiles.

In [15]:
df.describe()

account length  area code  number vmail messages  total day minutes  \
count         3333.00    3333.00                3333.00            3333.00   
mean           101.06     437.18                   8.10             179.78   
std             39.82      42.37                  13.69              54.47   
min              1.00     408.00                   0.00               0.00   
25%             74.00     408.00                   0.00             143.70   
50%            101.00     415.00                   0.00             179.40   
75%            127.00     510.00                  20.00             216.40   
max            243.00     510.00                  51.00             350.80   

       total day calls  total day charge  total eve minutes  total eve calls  \
count          3333.00           3333.00            3333.00          3333.00   
mean            100.44             30.56             200.98           100.11   
std              20.07              9.26              50.71            19.92   
min               0.00              0.00               0.00             0.00   
25%              87.00             24.43             166.60            87.00   
50%             101.00             30.50             201.40           100.00   
75%             114.00             36.79             235.30           114.00   
max             165.00             59.64             363.70           170.00   

       total eve charge  total night minutes  total night calls  \
count           3333.00              3333.00            3333.00   
mean              17.08               200.87             100.11   
std                4.31                50.57              19.57   
min                0.00                23.20              33.00   
25%               14.16               167.00              87.00   
50%               17.12               201.20             100.00   
75%               20.00               235.30             113.00   
max               30.91               395.00             175.00   

       total night charge  total intl minutes  total intl calls  \
count             3333.00             3333.00           3333.00   
mean                 9.04               10.24              4.48   
std                  2.28                2.79              2.46   
min                  1.04                0.00              0.00   
25%                  7.52                8.50              3.00   
50%                  9.05               10.30              4.00   
75%                 10.59               12.10              6.00   
max                 17.77               20.00             20.00   

       total intl charge  customer service calls    churn  
count            3333.00                 3333.00  3333.00  
mean                2.76                    1.56     0.14  
std                 0.75                    1.32     0.35  
min                 0.00                    0.00     0.00  
25%                 2.30                    1.00     0.00  
50%                 2.78                    1.00     0.00  
75%                 3.27                    2.00     0.00  
max                 5.40                    9.00     1.00

In order to see statistics on non-numerical features, one has to explicitly indicate data types of interest in the include parameter.

In [16]:
df.describe(include=["object", "bool"])

state phone number international plan voice mail plan
count   3333         3333               3333            3333
unique    51         3333                  2               2
top       WV     400-4344                 no              no
freq     106            1               3010            2411

For categorical (type object) and boolean (type bool) features we can use the value_counts method. Let's have a look at the distribution of Churn:

In [17]:
df["churn"].value_counts()

churn
0    2850
1     483
Name: count, dtype: int64

2850 users out of 3333 are loyal; their Churn value is 0. To calculate fractions, pass normalize=True to the value_counts function.

In [18]:
df["churn"].value_counts(normalize=True)

churn
0    0.86
1    0.14
Name: proportion, dtype: float64


### Sorting

A DataFrame can be sorted by the value of one of the variables (i.e columns). For example, we can sort by Total day charge (use ascending=False to sort in descending order):


In [19]:
df.sort_values(by="total day charge", ascending=False).head()

state  account length  area code phone number international plan  \
365     CO             154        415     343-5709                 no   
985     NY              64        415     345-9140                yes   
2594    OH             115        510     348-1163                yes   
156     OH              83        415     370-9116                 no   
605     MO             112        415     373-2053                 no   

     voice mail plan  number vmail messages  total day minutes  \
365               no                      0              350.8   
985               no                      0              346.8   
2594              no                      0              345.3   
156               no                      0              337.4   
605               no                      0              335.5   

      total day calls  total day charge  ...  total eve calls  \
365                75             59.64  ...               94   
985                55             58.96  ...               79   
2594               81             58.70  ...              106   
156               120             57.36  ...              116   
605                77             57.04  ...              109   

      total eve charge  total night minutes  total night calls  \
365              18.40                253.9                100   
985              21.21                275.4                102   
2594             17.29                217.5                107   
156              19.33                153.9                114   
605              18.06                265.0                132   

      total night charge  total intl minutes  total intl calls  \
365                11.43                10.1                 9   
985                12.39                13.3                 9   
2594                9.79                11.8                 8   
156                 6.93                15.8                 7   
605                11.93                12.7                 8   

      total intl charge  customer service calls  churn  
365                2.73                       1      1  
985                3.59                       1      1  
2594               3.19                       1      1  
156                4.27                       0      1  
605                3.43                       2      1  

[5 rows x 21 columns]

We can also sort by multiple columns:

In [20]:
df.sort_values(by=["churn", "total day charge"], ascending=[True, False]).head()

state  account length  area code phone number international plan  \
688     MN              13        510     338-7120                 no   
2259    NC             210        415     363-7802                 no   
534     LA              67        510     373-6784                 no   
575     SD             114        415     351-7369                 no   
2858    AL             141        510     388-8583                 no   

     voice mail plan  number vmail messages  total day minutes  \
688              yes                     21              315.6   
2259             yes                     31              313.8   
534               no                      0              310.4   
575              yes                     36              309.9   
2858             yes                     28              308.0   

      total day calls  total day charge  ...  total eve calls  \
688               105             53.65  ...               71   
2259               87             53.35  ...              103   
534                97             52.77  ...              123   
575                90             52.68  ...               89   
2858              123             52.36  ...              128   

      total eve charge  total night minutes  total night calls  \
688              17.76                260.1                123   
2259             12.55                192.7                 97   
534               5.65                246.5                 99   
575              17.03                183.5                105   
2858             21.06                152.9                103   

      total night charge  total intl minutes  total intl calls  \
688                11.70                12.1                 3   
2259                8.67                10.1                 7   
534                11.09                 9.2                10   
575                 8.26                14.2                 2   
2858                6.88                 7.4                 3   

      total intl charge  customer service calls  churn  
688                3.27                       3      0  
2259               2.73                       3      0  
534                2.48                       4      0  
575                3.83                       1      0  
2858               2.00                       1      0  

[5 rows x 21 columns]

### Indexing and retrieving data

A DataFrame can be indexed in a few different ways. 

To get a single column, you can use a DataFrame'Name' construction. Let's use this to answer a question about that column alone: what is the proportion of churned users in our dataframe?

In [21]:
df["churn"].mean()

np.float64(0.14491449144914492)

14.5% is actually quite bad for a company; such a churn rate can make the company go bankrupt.

Boolean indexing with one column is also very convenient. The syntax is df P(df['Name']), where P is some logical condition that is checked for each element of the Name column. The result of such indexing is the DataFrame consisting only of rows that satisfy the P condition on the Name column. 

Let's use it to answer the question:

What are the average values of numerical features for churned users?



In [22]:
df.select_dtypes(include=np.number)[df["churn"] == 1].mean()

account length            102.66
area code                 437.82
number vmail messages       5.12
total day minutes         206.91
total day calls           101.34
total day charge           35.18
total eve minutes         212.41
total eve calls           100.56
total eve charge           18.05
total night minutes       205.23
total night calls         100.40
total night charge          9.24
total intl minutes         10.70
total intl calls            4.16
total intl charge           2.89
customer service calls      2.23
churn                       1.00
dtype: float64

How much time (on average) do churned users spend on the phone during daytime?

In [23]:
df[df["churn"] == 1]["total day minutes"].mean()

np.float64(206.91407867494823)


What is the maximum length of international calls among loyal users (Churn == 0) who do not have an international plan?



In [24]:
df[(df["churn"] == 0) & (df["international plan"] == "No")]["total intl minutes"].max()

nan

DataFrames can be indexed by column name (label) or row name (index) or by the serial number of a row. The `loc` method is used for indexing by name, while `iloc()` is used for indexing by number.

In the first case below, we say "give us the values of the rows with index from 0 to 5 (inclusive) and columns labeled from State to Area code (inclusive)". In the second case, we say "give us the values of the first five rows in the first three columns" (as in a typical Python slice: the maximal value is not included).

In [25]:
df.loc[0:5, "state":"area code"]

state  account length  area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415
5    AL             118        510

In [26]:
df.iloc[0:5, 0:3]

state  account length  area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415

If we need the first or the last line of the data frame, we can use the df[:1] or df[-1:] construct:

In [27]:
df[-1:]

state  account length  area code phone number international plan  \
3332    TN              74        415     400-4344                 no   

     voice mail plan  number vmail messages  total day minutes  \
3332             yes                     25              234.4   

      total day calls  total day charge  ...  total eve calls  \
3332              113             39.85  ...               82   

      total eve charge  total night minutes  total night calls  \
3332              22.6                241.4                 77   

      total night charge  total intl minutes  total intl calls  \
3332               10.86                13.7                 4   

      total intl charge  customer service calls  churn  
3332                3.7                       0      0  

[1 rows x 21 columns]


### Applying Functions to Cells, Columns and Rows

To apply functions to each column, use apply():


In [28]:
df.apply(np.max)

state                           WY
account length                 243
area code                      510
phone number              422-9964
international plan             yes
voice mail plan                yes
number vmail messages           51
total day minutes            350.8
total day calls                165
total day charge             59.64
total eve minutes            363.7
total eve calls                170
total eve charge             30.91
total night minutes          395.0
total night calls              175
total night charge           17.77
total intl minutes            20.0
total intl calls                20
total intl charge              5.4
customer service calls           9
churn                            1
dtype: object

The apply method can also be used to apply a function to each row. To do this, specify axis=1. Lambda functions are very convenient in such scenarios. For example, if we need to select all states starting with W, we can do it like this:

In [29]:
df[df["state"].apply(lambda state: state[0] == "W")].head()

state  account length  area code phone number international plan  \
9     WV             141        415     330-8173                yes   
26    WY              57        408     357-3817                 no   
44    WI              64        510     352-1237                 no   
49    WY              97        415     405-7146                 no   
54    WY              87        415     353-3759                 no   

   voice mail plan  number vmail messages  total day minutes  total day calls  \
9              yes                     37              258.6               84   
26             yes                     39              213.0              115   
44              no                      0              154.0               67   
49             yes                     24              133.2              135   
54              no                      0              151.0               83   

    total day charge  ...  total eve calls  total eve charge  \
9              43.96  ...              111             18.87   
26             36.21  ...              112             16.24   
44             26.18  ...              118             19.19   
49             22.64  ...               58             18.46   
54             25.67  ...              116             18.67   

    total night minutes  total night calls  total night charge  \
9                 326.4                 97               14.69   
26                182.7                115                8.22   
44                265.3                 86               11.94   
49                 70.6                 79                3.18   
54                203.9                127                9.18   

    total intl minutes  total intl calls  total intl charge  \
9                 11.2                 5               3.02   
26                 9.5                 3               2.57   
44                 3.5                 3               0.95   
49                11.0                 3               2.97   
54                 9.7                 3               2.62   

    customer service calls  churn  
9                        0      0  
26                       0      0  
44                       1      0  
49                       1      0  
54                       5      1  

[5 rows x 21 columns]

The map method can be used to replace values in a column by passing a dictionary of the form {old_value: new_value} as its argument:

In [30]:
d = {"No": False, "Yes": True}
df["international plan"] = df["international plan"].map(d)
df.head()

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                NaN   
1    OH             107        415     371-7191                NaN   
2    NJ             137        415     358-1921                NaN   
3    OH              84        408     375-9999                NaN   
4    OK              75        415     330-6626                NaN   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0  

[5 rows x 21 columns]

The same thing can be done with the replace method:

In [31]:
df = df.replace({"voice mail plan": d})
df.head()

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                NaN   
1    OH             107        415     371-7191                NaN   
2    NJ             137        415     358-1921                NaN   
3    OH              84        408     375-9999                NaN   
4    OK              75        415     330-6626                NaN   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0  

[5 rows x 21 columns]


### Grouping

In general, grouping data in Pandas works as follows:




df.groupby(by=grouping_columns)[columns_to_show].function()



1. First, the groupby method divides the grouping_columns by their values. They become a new index in the resulting dataframe.
2. Then, columns of interest are selected (columns_to_show). If columns_to_show is not included, all non groupby clauses will be included.
3. Finally, one or several functions are applied to the obtained groups per selected columns.

Here is an example where we group the data according to the values of the Churn variable and display statistics of three columns in each group:

In [32]:
columns_to_show = ["total day minutes", "total eve minutes", "total night minutes"]

df.groupby(["churn"])[columns_to_show].describe(percentiles=[])

total day minutes                                   total eve minutes  \
                  count    mean    std  min    50%    max             count   
churn                                                                         
0                2850.0  175.18  50.18  0.0  177.2  315.6            2850.0   
1                 483.0  206.91  69.00  0.0  217.6  350.8             483.0   

                                         total night minutes                 \
         mean    std   min    50%    max               count    mean    std   
churn                                                                         
0      199.04  50.29   0.0  199.6  361.8              2850.0  200.13  51.11   
1      212.41  51.73  70.9  211.3  363.7               483.0  205.23  47.13   

                            
        min     50%    max  
churn                       
0      23.2  200.25  395.0  
1      47.4  204.80  354.9

Let’s do the same thing, but slightly differently by passing a list of functions to agg():

In [33]:
columns_to_show = ["total day minutes", "total eve minutes", "total night minutes"]

df.groupby(["churn"])[columns_to_show].agg([np.mean, np.std, np.min, np.max])

C:\Users\ASUS\AppData\Local\Temp\ipykernel_14300\2833449601.py:3: FutureWarning: The provided callable <function mean at 0x000002863262F740> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  df.groupby(["churn"])[columns_to_show].agg([np.mean, np.std, np.min, np.max])
C:\Users\ASUS\AppData\Local\Temp\ipykernel_14300\2833449601.py:3: FutureWarning: The provided callable <function std at 0x000002863262F880> is currently using SeriesGroupBy.std. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "std" instead.
  df.groupby(["churn"])[columns_to_show].agg([np.mean, np.std, np.min, np.max])
C:\Users\ASUS\AppData\Local\Temp\ipykernel_14300\2833449601.py:3: FutureWarning: The provided callable <function min at 0x000002863262EE80> is currently using SeriesGroupBy.min. In a future version of pandas, the prov

total day minutes                    total eve minutes               \
                   mean    std  min    max              mean    std   min   
churn                                                                       
0                175.18  50.18  0.0  315.6            199.04  50.29   0.0   
1                206.91  69.00  0.0  350.8            212.41  51.73  70.9   

             total night minutes                      
         max                mean    std   min    max  
churn                                                 
0      361.8              200.13  51.11  23.2  395.0  
1      363.7              205.23  47.13  47.4  354.9


### Summary tables

Suppose we want to see how the observations in our sample are distributed in the context of two variables - Churn and International plan. To do so, we can build a contingency table using the crosstab method:



In [34]:
pd.crosstab(df["churn"], df["voice mail plan"], normalize=True)

voice mail plan    no   yes
churn                      
0                0.60  0.25
1                0.12  0.02

We can see that most of the users are loyal and do not use additional services (International Plan/Voice mail).

This will resemble pivot tables to those familiar with Excel. And, of course, pivot tables are implemented in Pandas: the pivot_table method takes the following parameters:

* values – a list of variables to calculate statistics for,
* index – a list of variables to group data by,
* aggfunc – what statistics we need to calculate for groups, ex. sum, mean, maximum, minimum or something else.

Let's take a look at the average number of day, evening, and night calls by area code:

In [35]:
df.pivot_table(
    ["total day calls", "total eve calls", "total night calls"],
    ["area code"],
    aggfunc="mean",
)

total day calls  total eve calls  total night calls
area code                                                     
408                 100.50            99.79              99.04
415                 100.58           100.50             100.40
510                 100.10            99.67             100.60


### DataFrame transformations

Like many other things in Pandas, adding columns to a DataFrame is doable in many ways.

For example, if we want to calculate the total number of calls for all users, let's create the total_calls Series and paste it into the DataFrame:



In [36]:
total_calls = (
    df["total day calls"]
    + df["total eve calls"]
    + df["total night calls"]
    + df["total intl calls"]
)
df.insert(loc=len(df.columns), column="total calls", value=total_calls)

df.head()

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                NaN   
1    OH             107        415     371-7191                NaN   
2    NJ             137        415     358-1921                NaN   
3    OH              84        408     375-9999                NaN   
4    OK              75        415     330-6626                NaN   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...  total eve charge  total night minutes  \
0             45.07  ...             16.78                244.7   
1             27.47  ...             16.62                254.4   
2             41.38  ...             10.30                162.6   
3             50.90  ...              5.26                196.9   
4             28.34  ...             12.61                186.9   

   total night calls  total night charge  total intl minutes  \
0                 91               11.01                10.0   
1                103               11.45                13.7   
2                104                7.32                12.2   
3                 89                8.86                 6.6   
4                121                8.41                10.1   

   total intl calls  total intl charge  customer service calls  churn  \
0                 3               2.70                       1      0   
1                 3               3.70                       1      0   
2                 5               3.29                       0      0   
3                 7               1.78                       2      0   
4                 3               2.73                       3      0   

   total calls  
0          303  
1          332  
2          333  
3          255  
4          359  

[5 rows x 22 columns]

It is possible to add a column more easily without creating an intermediate Series instance:

In [37]:
df["total charge"] = (
    df["total day charge"]
    + df["total eve charge"]
    + df["total night charge"]
    + df["total intl charge"]
)
df.head()

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                NaN   
1    OH             107        415     371-7191                NaN   
2    NJ             137        415     358-1921                NaN   
3    OH              84        408     375-9999                NaN   
4    OK              75        415     330-6626                NaN   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...  total night minutes  total night calls  \
0             45.07  ...                244.7                 91   
1             27.47  ...                254.4                103   
2             41.38  ...                162.6                104   
3             50.90  ...                196.9                 89   
4             28.34  ...                186.9                121   

   total night charge  total intl minutes  total intl calls  \
0               11.01                10.0                 3   
1               11.45                13.7                 3   
2                7.32                12.2                 5   
3                8.86                 6.6                 7   
4                8.41                10.1                 3   

   total intl charge  customer service calls  churn  total calls  total charge  
0               2.70                       1      0          303         75.56  
1               3.70                       1      0          332         59.24  
2               3.29                       0      0          333         62.29  
3               1.78                       2      0          255         66.80  
4               2.73                       3      0          359         52.09  

[5 rows x 23 columns]

To delete columns or rows, use the drop method, passing the required indexes and the axis parameter (1 if you delete columns, and nothing or 0 if you delete rows). The inplace argument tells whether to change the original DataFrame. With inplace=False, the drop method doesn't change the existing DataFrame and returns a new one with dropped rows or columns. With inplace=True, it alters the DataFrame.

In [38]:
df.drop(["total charge", "total calls"], axis=1, inplace=True)
df.drop([1, 2]).head()

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                NaN   
3    OH              84        408     375-9999                NaN   
4    OK              75        415     330-6626                NaN   
5    AL             118        510     391-8027                NaN   
6    MA             121        510     355-9993                NaN   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
3              no                      0              299.4               71   
4              no                      0              166.7              113   
5              no                      0              223.4               98   
6             yes                     24              218.2               88   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   
5             37.98  ...              101             18.75   
6             37.09  ...              108             29.62   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
3                196.9                 89                8.86   
4                186.9                121                8.41   
5                203.9                118                9.18   
6                212.6                118                9.57   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   
5                 6.3                 6               1.70   
6                 7.5                 7               2.03   

   customer service calls  churn  
0                       1      0  
3                       2      0  
4                       3      0  
5                       0      0  
6                       3      0  

[5 rows x 21 columns]

## 2. First attempt at predicting telecom churn


Let's see how churn rate is related to the *International plan* feature. We'll do this using a crosstab contingency table and also through visual analysis with Seaborn (however, visual analysis will be covered more thoroughly in the next article).


In [42]:
import matplotlib.pyplot as plt
import seaborn as sns

%config InlineBackend.figure_format = 'retina'

ModuleNotFoundError: No module named 'matplotlib'


We see that, with International Plan, the churn rate is much higher, which is an interesting observation! Perhaps large and poorly controlled expenses with international calls are very conflict-prone and lead to dissatisfaction among the telecom operator's customers.

Next, let's look at another important feature – Customer service calls. Let's also make a summary table and a picture.

In [43]:
pd.crosstab(df["churn"], df["customer service calls"], margins=True)

customer service calls    0     1    2    3    4   5   6  7  8  9   All
churn                                                                  
0                       605  1059  672  385   90  26   8  4  1  0  2850
1                        92   122   87   44   76  40  14  5  1  2   483
All                     697  1181  759  429  166  66  22  9  2  2  3333

In [44]:
sns.countplot(x="customer service calls", hue="churn", data=df);

NameError: name 'sns' is not defined

Although it's not so obvious from the summary table, it's easy to see from the above plot that the churn rate increases sharply from 4 customer service calls and above.

Now let's add a binary feature to our DataFrame – Customer service calls > 3. And once again, let's see how it relates to churn. 

In [45]:
df["many_service_calls"] = (df["customer service calls"] > 3).astype("int")

pd.crosstab(df["many_service_calls"], df["churn"], margins=True)

churn                  0    1   All
many_service_calls                 
0                   2721  345  3066
1                    129  138   267
All                 2850  483  3333

In [46]:
sns.countplot(x="many_service_calls", hue="churn", data=df);

NameError: name 'sns' is not defined


Let's construct another contingency table that relates Churn with both International plan and freshly created Many_service_calls.



In [47]:
pd.crosstab(df["many_service_calls"] & df["international plan"], df["churn"])

churn     0    1
row_0           
False  2850  483

Thus, by predicting that a customer is not loyal (Churn=1) in the case when the number of calls to the service center is greater than 3 and the International Plan is added (and predicting Churn=0 otherwise), we might expect an accuracy of 85.8% (we are mistaken only 464 + 9 times). This number, 85.8%, that we got through this very simple reasoning serves as a good starting point (baseline) for the development of future machine learning models.

As we move on in this course, recall that, before the advent of machine learning, the data analysis process looked something like this. Let's recap what we've covered:
    
* The share of loyal clients in the sample is 85.5%. The most naive model that always predicts a "loyal customer" on such data will guess right in about 85.5% of all cases. That is, the proportion of correct answers (accuracy) of subsequent models should be no less than this number, and will hopefully be significantly higher;